### PR TITLE
[FEATURE] Enable Simultaneous Execution of all Metric Computations for ParameterBuilder implementations in Rule-Based Profiler

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
Enable computation of all metric configurations for all batches in a single method call and parse returned results accordingly.

As a strategy, since "metric_domain_kwargs" changes depending on "batch_id", "metric_value_kwargs" serves as
identifying entity (through "AttributedResolvedMetrics") for accessing resolved metrics (computation results).

All "MetricConfiguration" directives are generated by combining each metric_value_kwargs" with
"metric_domain_kwargs" for all "batch_ids" (where every "metric_domain_kwargs" represents separate "batch_id").

Then, all "MetricConfiguration" objects, collected into list as container, are resolved simultaneously.

All available ParameterBuilder implementations (including `SimpleDateFormatStringParameterBuilder` and `RegexPatternStringParameterBuilder`, which exercise the new functionality) have been updated accordingly.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: GREAT-584
-
-

After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py [hunk 1]
```diff
@@ -34,6 +37,35 @@
...
+    """
+
+    metric_values: MetricValues
+    metric_attributes: Attributes
+
+    def add_resolved_metric(self, value: Any) -> None:
+        if self.metric_values is None:
+            self.metric_values = []
+
+        self.metric_values.append(value)
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:56]: Can it really be `Any`? If the attribute is of type `MetricValues`, can we narrow this down?

**Review by NathanFarmer — APPROVED** [review]
Not able to find a single thing to even nitpick. LGTM!

File: great_expectations/rule_based_profiler/parameter_builder/regex_pattern_string_parameter_builder.py [hunk 2]
```diff
@@ -101,12 +101,7 @@ def threshold(self) -> Union[str, float]:
     @property
     def candidate_regexes(
         self,
-    ) -> Union[
-        str,
-        Union[
-            Set[str], List[str], "RegexPatternStringParameterBuilder.CANDIDATE_REGEX"
-        ],
-    ]:  # noqa: F821
+    ) -> Union[str, Union[List[str], Set[str],],]:
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/parameter_builder/regex_pattern_string_parameter_builder.py]: A little confused by this type annotation - should the commas be here?

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py [hunk 3]
```diff
@@ -207,26 +243,80 @@ def get_metrics(
...
+            # Obtain value kwargs from "rule state" (i.e., variables and parameters); from instance variable otherwise.
+            get_parameter_value_and_validate_return_type(
+                domain=domain,
+                parameter_reference=value_kwargs_cursor,
+                expected_return_type=None,
+                variables=variables,
+                parameters=parameters,
+            )
+            for value_kwargs_cursor in metric_value_kwargs
+        ]
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:297]: You were right about me not loving the comprehensions 😬 

I feel like a loop is more readable here but this is just nitpicking. Use what you think is best

**Review by cdkini — APPROVED** [review]

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:56]: @cdkini Yes, it is a computed metric -- truly anything (string, list, number of all kinds, etc.)!

File: great_expectations/rule_based_profiler/parameter_builder/regex_pattern_string_parameter_builder.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/regex_pattern_string_parameter_builder.py]: @cdkini That was there before, I just got rid of the unnecessary (and incorrect) quoted type and simplified it.  I got rid of the commas (and `black` seems to be OK with it -- in next commit).

**Pull request merged**